### PR TITLE
CDAP-3383 Enable tx pruning in CDAP

### PR DESCRIPTION
**Pull request by gokulavasan**
JIRA : https://issues.cask.co/browse/CDAP-3383

Build : http://builds.cask.co/browse/CDAP-DUT5271

### CODE REVIEW

**Comment by gokulavasan** [comment]
Verified on a cluster. @poornachandra Please review when you get a chance.

File: cdap-hbase-compat-1.0/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase10/MessageTableRegionObserver.java [hunk 1]
```diff
@@ -143,12 +142,12 @@ private TopicMetadataCache getTopicMetadataCache(RegionCoprocessorEnvironment en
...
 
   private TopicMetadataCache createTopicMetadataCache(RegionCoprocessorEnvironment env) {
-    return new TopicMetadataCache(env, cConfReader, nameConverter, hbaseNamespacePrefix, metadataTableNamespace,
+    return new TopicMetadataCache(env, cConfReader, hbaseNamespacePrefix, metadataTableNamespace,
                                   new HBase10ScanBuilder());
   }
 
   private Supplier<TransactionStateCache> getTransactionStateCacheSupplier(String tablePrefix, Configuration conf) {
-    String sysConfigTablePrefix = nameConverter.getSysConfigTablePrefix(tablePrefix);
+    String sysConfigTablePrefix = HTableNameConverter.getSysConfigTablePrefix(tablePrefix);
```
Comment (poornachandra) [inline cdap-hbase-compat-1.0/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase10/MessageTableRegionObserver.java:191]: Unrelated change?

File: cdap-hbase-compat-1.1/src/main/java/co/cask/data2/txprune/DefaultHBaseTransactionPruningPlugin.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * {@link TransactionPruningPlugin} for CDAP Datasets, TMS Message Table, HBase Queues.
+ */
+public class DefaultHBaseTransactionPruningPlugin extends HBaseTransactionPruningPlugin {
+
+  @Override
+  protected boolean isTransactionalTable(HTableDescriptor tableDescriptor) {
+    return tableDescriptor.hasCoprocessor(DefaultTransactionProcessor.class.getName()) ||
+      tableDescriptor.hasCoprocessor(MessageTableRegionObserver.class.getName()) ||
+      tableDescriptor.hasCoprocessor(HBaseQueueRegionObserver.class.getName());
```
Comment (poornachandra) [inline cdap-hbase-compat-1.1/src/main/java/co/cask/data2/txprune/DefaultHBaseTransactionPruningPlugin.java:35]: What about `IncrementHandler` and `PayloadTableRegionObserver`?

File: cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/coprocessor/hbase11/DefaultTransactionProcessor.java [hunk 3]
```diff
@@ -16,31 +16,136 @@
...
+                                       @Nullable Transaction tx) throws IOException {
+    if (tx == null) {
+      return;
+    }
+
+    long maxLifetimeMillis;
+    if (txMaxLifetimeMillis == null) {
+      Configuration conf = getConfiguration(env);
+      if (conf != null) {
+        this.txMaxLifetimeMillis = TimeUnit.SECONDS.toMillis(conf.getInt(TxConstants.Manager.CFG_TX_MAX_LIFETIME,
```
Comment (poornachandra) [inline cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/coprocessor/hbase11/DefaultTransactionProcessor.java:95]: Looks like CDAP allows configuration to be loaded dynamically when CDAP is running. If so, we'll need to keep the same behavior here too.

**Review by poornachandra — COMMENTED** [review]
@gokulavasan A few comments to address

File: cdap-hbase-compat-1.0/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase10/MessageTableRegionObserver.java (hunk 1 above)
Comment (gokulavasan) [reply cdap-hbase-compat-1.0/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase10/MessageTableRegionObserver.java:191]: Yes it is unrelated but I thought it would be good to make that change to use the static method.

File: cdap-hbase-compat-1.1/src/main/java/co/cask/data2/txprune/DefaultHBaseTransactionPruningPlugin.java (hunk 2 above)
Comment (gokulavasan) [reply cdap-hbase-compat-1.1/src/main/java/co/cask/data2/txprune/DefaultHBaseTransactionPruningPlugin.java:35]: PayloadTableRegionObserver doesn't clean up messages based on invalid transactions. It will clean up only based on TTL and deleted topics.

Comment (gokulavasan) [inline cdap-hbase-compat-1.1/src/main/java/co/cask/data2/txprune/DefaultHBaseTransactionPruningPlugin.java]: For IncrementHandler, I don't see it using the TransactionVisibilitySnapshot. I will dig a little more into it.

File: cdap-hbase-compat-1.0/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase10/MessageTableRegionObserver.java (hunk 1 above)
Comment (poornachandra) [reply cdap-hbase-compat-1.0/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase10/MessageTableRegionObserver.java:191]: cool

File: cdap-hbase-compat-1.1/src/main/java/co/cask/data2/txprune/DefaultHBaseTransactionPruningPlugin.java (hunk 2 above)
Comment (gokulavasan) [reply cdap-hbase-compat-1.1/src/main/java/co/cask/data2/txprune/DefaultHBaseTransactionPruningPlugin.java:35]: For IncrementHandler, it is always used in conjunction with DataJanitor (DefaultTransactionProcessor) and the DataJanitor gets higher priority over IncrementHandler when it is present. I will add comments at the relevant code regarding this.

File: cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/coprocessor/hbase11/DefaultTransactionProcessor.java (hunk 3 above)
Comment (gokulavasan) [reply cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/coprocessor/hbase11/DefaultTransactionProcessor.java:95]: Sure yes, I will make that change.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java [hunk 4]
```diff
@@ -92,6 +94,9 @@
   // name length + name of the table: handy to have one cached
   private final byte[] nameAsTxChangePrefix;
   private final boolean safeReadlessIncrements;
+  // tx max lifetime property comes usually from cConf in DefaultTransactionProcessor but if it is not available
+  // briefly during startup, fetch the coprocessor gets it from the operation's attribute.
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java]: `during startup, fetch the coprocessor` => `fetch` is not required

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/coprocessor/ConsumerConfigCache.java [hunk 5]
```diff
@@ -105,6 +105,11 @@ public boolean isAlive() {
   }
 
   @Nullable
+  public CConfiguration getCConf() {
+    return (conf != null) ? CConfiguration.copy(conf) : null;
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/coprocessor/ConsumerConfigCache.java:109]: Since `conf` will be accessed by multiple threads, it is good to make it volatile

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * {@link Thread} that refreshes {@link CConfiguration} periodically.
+ */
+public class CConfigurationCache {
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java:33]: It would be good to make this singleton, so that a single instance can be shared across all the co-processors for various regions in a region server.

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(CConfigurationCache.class);
+  private static final String CCONF_UPDATE_PERIOD = "cdap.transaction.coprocessor.configuration.update.period.secs";
+  private static final Long DEFAULT_CCONF_UPDATE_PERIOD = TimeUnit.MINUTES.toMillis(5);
+
+
+  private final CConfigurationReader cConfReader;
+
+  private Thread refreshThread;
+  private CConfiguration cConf;
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java]: This has to be volatile too

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java (hunk 6 above)
Comment (poornachandra) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java:33]: We should do the same in Tephra too

File: cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase11/MessageTableRegionObserver.java [hunk 8]
```diff
@@ -128,6 +135,28 @@ public InternalScanner preCompactScannerOpen(ObserverContext<RegionCoprocessorEn
...
+      CConfiguration cConf = metadataCache.getCConfiguration();
+      if (cConf != null) {
+        pruneEnable = cConf.getBoolean(TxConstants.TransactionPruning.PRUNE_ENABLE,
+                                       TxConstants.TransactionPruning.DEFAULT_PRUNE_ENABLE);
+        String pruneTable = cConf.get(TxConstants.TransactionPruning.PRUNE_STATE_TABLE,
+                                      TxConstants.TransactionPruning.DEFAULT_PRUNE_STATE_TABLE);
+        long pruneFlushInterval = TimeUnit.SECONDS.toMillis(
+          cConf.getLong(TxConstants.TransactionPruning.PRUNE_FLUSH_INTERVAL,
+                        TxConstants.TransactionPruning.DEFAULT_PRUNE_FLUSH_INTERVAL));
+        compactionState = new CompactionState(c.getEnvironment(), TableName.valueOf(pruneTable), pruneFlushInterval);
```
Comment (poornachandra) [inline cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase11/MessageTableRegionObserver.java]: We need to create compaction state only when `pruneEnable` is true

File: cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/queue/coprocessor/hbase11/HBaseQueueRegionObserver.java [hunk 9]
```diff
@@ -137,7 +145,40 @@ public InternalScanner preCompact(ObserverContext<RegionCoprocessorEnvironment>
...
+      CConfiguration cConf = consumerConfigCache.getCConf();
+      if (cConf != null) {
+        pruneEnable = cConf.getBoolean(TxConstants.TransactionPruning.PRUNE_ENABLE,
+                                       TxConstants.TransactionPruning.DEFAULT_PRUNE_ENABLE);
+        String pruneTable = cConf.get(TxConstants.TransactionPruning.PRUNE_STATE_TABLE,
+                                      TxConstants.TransactionPruning.DEFAULT_PRUNE_STATE_TABLE);
+        long pruneFlushInterval = TimeUnit.SECONDS.toMillis(
+          cConf.getLong(TxConstants.TransactionPruning.PRUNE_FLUSH_INTERVAL,
+                        TxConstants.TransactionPruning.DEFAULT_PRUNE_FLUSH_INTERVAL));
+        compactionState = new CompactionState(e.getEnvironment(), TableName.valueOf(pruneTable), pruneFlushInterval);
```
Comment (poornachandra) [inline cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/queue/coprocessor/hbase11/HBaseQueueRegionObserver.java]: Same here about creating this only if `pruneEnable` is true

File: cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/queue/coprocessor/hbase11/HBaseQueueRegionObserver.java [hunk 10]
```diff
@@ -188,11 +230,14 @@ private TransactionStateCache getTxStateCache() {
     private long rowsEvicted = 0;
     // couldn't be evicted due to incomplete view of row
     private long skippedIncomplete = 0;
+    private boolean invalidTxData;
 
-    private EvictionInternalScanner(String action, RegionCoprocessorEnvironment env, InternalScanner scanner) {
+    private EvictionInternalScanner(String action, RegionCoprocessorEnvironment env, InternalScanner scanner,
+                                    @Nullable TransactionVisibilityState state) {
```
Comment (poornachandra) [inline cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/queue/coprocessor/hbase11/HBaseQueueRegionObserver.java:238]: Can `TransactionVisibilityState` be null? If so we should not record compaction state in such cases, right?

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * {@link Thread} that refreshes {@link CConfiguration} periodically.
+ */
+public class CConfigurationCache {
+  private static final Logger LOG = LoggerFactory.getLogger(CConfigurationCache.class);
+  private static final String CCONF_UPDATE_PERIOD = "cdap.transaction.coprocessor.configuration.update.period.secs";
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java:35]: Ideally, we should have one config for config update interval

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/messaging/TopicMetadataCache.java [hunk 12]
```diff
@@ -57,23 +57,22 @@
...
   private final String metadataTableNamespace;
   private final ScanBuilder scanBuilder;
 
   private Thread refreshThread;
   private long lastUpdated;
+
   private volatile Map<ByteBuffer, Map<String, String>> metadataCache = new HashMap<>();
   private long metadataCacheUpdateFreqInMillis = TimeUnit.SECONDS.toMillis(
     MessagingUtils.Constants.METADATA_CACHE_UPDATE_FREQUENCY_SECS);
+  private CConfiguration cConf;
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/messaging/TopicMetadataCache.java]: `cConf` has to be volatile here too

**Review by poornachandra — COMMENTED** [review]
@gokulavasan A few more comments

File: cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/queue/coprocessor/hbase11/HBaseQueueRegionObserver.java (hunk 10 above)
Comment (gokulavasan) [reply cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/queue/coprocessor/hbase11/HBaseQueueRegionObserver.java:238]: CompactionState#record has TransactionVisibilityState to be ``@Nullable``

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java (hunk 11 above)
Comment (gokulavasan) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java:35]: We seem to have one more in QueueConstants. We should deprecate that and I can make this more generic configuration property.

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java (hunk 6 above)
Comment (gokulavasan) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java:33]: I didn't quite understand the Tephra reference. Do we have a configuration refresh thread in Tephra?

Comment (gokulavasan) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java:33]: Also, would it be ok if we address the singleton issue in a separate JIRA?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/coprocessor/ConsumerConfigCache.java (hunk 5 above)
Comment (gokulavasan) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/coprocessor/ConsumerConfigCache.java:109]: Done

File: cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/queue/coprocessor/hbase11/HBaseQueueRegionObserver.java (hunk 10 above)
Comment (poornachandra) [reply cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/transaction/queue/coprocessor/hbase11/HBaseQueueRegionObserver.java:238]: Just checked that we handle null `TransactionVisibilityState` in `CompactionState`. In that case we don't write compaction state.

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java (hunk 6 above)
Comment (poornachandra) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java:33]: I am talking about `PruneUpperBoundWriter`. I think it is okay to address this in a separate JIRA

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java (hunk 11 above)
Comment (poornachandra) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java:35]: File a JIRA for this

**Comment by poornachandra** [comment]
@gokulavasan Please file the 3 JIRAs and add comment about IncrementHandler. The changes look good.

File: cdap-hbase-compat-1.2-cdh5.7.0/src/main/java/co/cask/data2/txprune/DefaultHBaseTransactionPruningPlugin.java [hunk 13]
```diff
@@ -19,16 +19,43 @@
...
 
+import java.io.IOException;
+
 /**
  * {@link TransactionPruningPlugin} for CDAP Datasets, TMS Message Table, HBase Queues.
  */
 public class DefaultHBaseTransactionPruningPlugin extends HBaseTransactionPruningPlugin {
 
   @Override
+  protected void createPruneTable(TableName stateTable) throws IOException {
```
Comment (poornachandra) [inline cdap-hbase-compat-1.2-cdh5.7.0/src/main/java/co/cask/data2/txprune/DefaultHBaseTransactionPruningPlugin.java]: `HBaseTransactionPruningPlugin` should have created the table, right? Do we have to create the table in this plugin too? Also I don't see this in all compat modules.

**Review by poornachandra — COMMENTED** [review]
@gokulavasan Just one comment for now

File: cdap-hbase-compat-1.2-cdh5.7.0/src/main/java/co/cask/data2/txprune/DefaultHBaseTransactionPruningPlugin.java (hunk 13 above)
Comment (gokulavasan) [reply cdap-hbase-compat-1.2-cdh5.7.0/src/main/java/co/cask/data2/txprune/DefaultHBaseTransactionPruningPlugin.java]: This is because there is a API difference between hbase-1.2-cdh5.7.0. HTableDescriptor#addFamily returns null in this version while it returns HTableDescriptor in the HBase 1.1 tephra compat which is what this version depends on.

Comment (poornachandra) [reply cdap-hbase-compat-1.2-cdh5.7.0/src/main/java/co/cask/data2/txprune/DefaultHBaseTransactionPruningPlugin.java]: I see, can you add a comment then?

**Comment by poornachandra** [comment]
@gokulavasan Changes look good, please file the 3 JIRAs and add the two comments - one in IncrementHandler and other in DefaultHBaseTransactionPruningPlugin for hbase-1.2-cdh5.7.0

**Comment by gokulavasan** [comment]
@poornachandra Thanks for the review Poorna. Opened https://issues.apache.org/jira/browse/TEPHRA-215, https://issues.cask.co/browse/CDAP-8324, https://issues.cask.co/browse/CDAP-8323 and also added comments. I will squash commits and wait for the build to complete.

**Comment by gokulavasan** [comment]
Build is 🍏 . Merging
